src/go/configgenerator/listener_generator_test.go

```diff
@@ -1472,6 +1472,101 @@ func TestPathMatcherFilter(t *testing.T) {
 	}
 }
 
+func TestServiceControl(t *testing.T) {
+	fakeServiceConfig := &confpb.Service{
+		Name: testProjectName,
+		Apis: []*apipb.Api{
+			{
+				Name: testApiName,
+				Methods: []*apipb.Method{
+					{
+						Name: "ListShelves",
+					},
+				},
+			},
+		},
+		Control: &confpb.Control{
+			Environment: statPrefix,
+		},
+	}
+	testData := []struct {
+		desc                            string
+		serviceControlCredentials       *options.IAMCredentialsOptions
+		serviceAccountKey               string
+		wantPartialServiceControlFilter string
+	}{
+		{
+			desc: "get access token from imds",
+			wantPartialServiceControlFilter: `
+    "imdsToken": {
+      "cluster": "metadata-cluster",
+      "timeout": "30s",
+      "uri": "http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token"
+    },`,
+		},
+		{
+			desc: "get access token from iam",
+			serviceControlCredentials: &options.IAMCredentialsOptions{
+				ServiceAccountEmail: "ServiceControl@iam.com",
+				Delegates:           []string{"delegate_foo", "delegate_bar"},
+			},
+			wantPartialServiceControlFilter: `
+    "iamToken": {
+      "accessToken": {
+        "remoteToken": {
+          "cluster": "metadata-cluster",
+          "timeout": "30s",
+          "uri": "http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token"
+        }
+      },
+      "delegates": [
+        "delegate_foo",
+        "delegate_bar"
+      ],
+      "iamUri": {
+        "cluster": "iam-cluster",
+        "timeout": "30s",
+        "uri": "https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/ServiceControl@iam.com:generateAccessToken"
+      },
+      "serviceAccountEmail": "ServiceControl@iam.com"
+    },`,
+		},
+		{
+			desc:              "get access token from the local access token server",
+			serviceAccountKey: "this-is-sa-cred",
+			wantPartialServiceControlFilter: `
+    "imdsToken": {
+      "cluster": "local-access-token-server-cluster",
+      "timeout": "30s",
+      "uri": "http://127.0.0.1:8791/v1/instance/service-accounts/default/token"
```

Comment:
Nit: The URL here does not match the IMDS url, it's missing `/computeMetadata/`

```
http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/token
```

Probably doesn't matter. But should we keep them the same for consistency?